router/data_store.go

```diff
@@ -39,196 +33,315 @@ type SyncHandler interface {
 	Remove(id string) error
 }
 
-func NewEtcdDataStore(etcd EtcdClient, prefix string) DataStore {
-	return &etcdDataStore{
-		prefix:   prefix,
-		etcd:     etcd,
-		stopSync: make(chan bool),
-	}
-}
+type pgDataStore struct {
+	pgx *pgx.ConnPool
 
-type etcdDataStore struct {
-	prefix   string
-	etcd     EtcdClient
-	stopSync chan bool
-}
+	routeType string
+	tableName string
 
-var ErrExists = errors.New("router: route already exists")
-var ErrNotFound = errors.New("router: route not found")
+	doneo sync.Once
+	donec chan struct{}
+}
 
-func (s *etcdDataStore) Add(r *router.Route) error {
-	data, err := json.Marshal(r)
-	if err != nil {
-		return err
+// NewPostgresDataStore returns a DataStore that stores route information in a
+// Postgres database. It uses pg_notify and a listener connection to watch for
+// route changes.
+func NewPostgresDataStore(routeType string, pgx *pgx.ConnPool) *pgDataStore {
+	tableName := ""
+	switch routeType {
+	case "http":
+		tableName = "http_routes"
+	case "tcp":
+		tableName = "tcp_routes"
+	default:
+		panic(fmt.Sprintf("unknown routeType: %q", routeType))
 	}
-	_, err = s.etcd.Create(s.path(r.ID), string(data), 0)
-	if e, ok := err.(*etcd.EtcdError); ok && e.ErrorCode == 105 {
-		err = ErrExists
+	return &pgDataStore{
+		pgx:       pgx,
+		routeType: routeType,
+		tableName: tableName,
+		donec:     make(chan struct{}),
 	}
-	return err
 }
 
-func (s *etcdDataStore) Set(r *router.Route) error {
-	data, err := json.Marshal(r)
-	if err != nil {
-		return err
+const sqlAddRouteHTTP = `
+INSERT INTO %s (parent_ref, service, domain, tls_cert, tls_key, sticky)
+	VALUES ($1, $2, $3, $4, $5, $6)
+	RETURNING id, created_at, updated_at`
+
+const sqlAddRouteTCP = `
+INSERT INTO %s (parent_ref, service, port)
+	VALUES ($1, $2, $3)
+	RETURNING id, created_at, updated_at`
+
+func (d *pgDataStore) Add(r *router.Route) (err error) {
+	switch d.tableName {
+	case "http_routes":
+		err = d.pgx.QueryRow(
+			fmt.Sprintf(sqlAddRouteHTTP, d.tableName),
```

Comment:
Something like this would at least avoid redefining that string in multiple places:

```go
const sqlAddRouteHTTP = `
INSERT INTO ` + tableNameHTTP + ` (parent_ref, service, domain, tls_cert, tls_key, sticky)
	VALUES ($1, $2, $3, $4, $5, $6)
	RETURNING id, created_at, updated_at`

const sqlAddRouteTCP = `
INSERT INTO ` + tableNameTCP + ` (parent_ref, service, port)
	VALUES ($1, $2, $3)
	RETURNING id, created_at, updated_at`
```

that ok?